### PR TITLE
[Design2-129] DSCO - Colors - Add ThemeContext

### PR DESCRIPTION
**Pull request by levadadenys**
### [Design2-129] DSCO - Colors - Add ThemeContext

* added ThemeContext.tsx

Upon working on Design System Color variables I investigated the possibility of adding support of switching Theme, which with this soultion turns out to be pretty clean and simple in my opinion. We haven't yet designed the switch themes user flow, but once it'll be designed and we have all the DSCO colors ready for dev - we'll be ready to implement it!

This solution supports using one theme in scope of the other theme (by nesting ThemeProviders).

## Links

- jira ticket: [Design2-129](https://codedotorg.atlassian.net/browse/DESIGN2-129)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/componentLibrary/common/contexts/ThemeContext.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+export const ThemeProvider = ({children}: {children: ReactNode}) => {
+  const [theme, setTheme] = useState<Theme>('Light');
+
+  const toggleTheme = () => {
+    setTheme(prevTheme => (prevTheme === 'Light' ? 'Dark' : 'Light'));
+  };
+
+  return (
+    <ThemeContext.Provider value={{theme, toggleTheme, setTheme}}>
+      <div data-theme={theme}>{children}</div>
```
Comment (sanchitmalhotra126) [inline apps/src/componentLibrary/common/contexts/ThemeContext.tsx:66]: ~~Curious why we're using the `data-theme` attribute? I'm assuming children would just grab the value out of the context directly, but is there another use for this?~~ Edit: makes more sense after seeing https://github.com/code-dot-org/code-dot-org/pull/59609 - nice solution!

**Review by sanchitmalhotra126 — COMMENTED** [review]
Nice! One question but otherwise this approach looks good to me

File: apps/src/componentLibrary/common/contexts/ThemeContext.tsx (hunk 1 above)
Comment (levadadenys) [reply apps/src/componentLibrary/common/contexts/ThemeContext.tsx:66]: Thanks for bringing it up! Added an additional comment just before return statement explaining this

File: apps/src/componentLibrary/common/contexts/ThemeContext.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  theme: Theme;
+  /** Toggle between themes (light/dark)*/
+  toggleTheme: () => void;
+  /** In case we'll have more than 2 themes - we'll be able to update theme value via this callback */
+  setTheme: (theme: Theme) => void;
+}
+
+/**
+ * Handles multiple themes logic.
+ * Can ThemeProvider can be stacked, meaning that you can have different theme in existing themeContext by
```
Comment (kelbyhawn) [inline apps/src/componentLibrary/common/contexts/ThemeContext.tsx]: Minor update here:
```suggestion
 *ThemeProvider can be stacked, meaning that you can have different theme in existing themeContext by
```

**Review by kelbyhawn — APPROVED** [review]
One small typo on a comment, but otherwise looks great! 🎉

**Review by sanchitmalhotra126 — APPROVED** [review]

**Comment by levadadenys** [comment]
Merging as is for now since for now it won’t be used anyways, there’re no user flows and designs for this functionality yet.
Though once we're ready - come back to it.

Upon further discussion there's a question to be solved in the future. Since we don't have one react entry point (e.g. `App.tsx`) and routing is handled by rails - we'll need to wrap every page's top level React component with ThemeContext, and also we'll need to investigate and decide optimal solution on whether to save theme value to the localStorage(keep theme preferences by browser) or to the db(backend endpoint) (keep theme preferences by user)

https://codedotorg.slack.com/archives/C043ZKQ4BS6/p1720458134634489

**Pull request merged**